File: internal/model/indexed_knowledge_item.go
package model

import (
	"time"

	"github.com/usesnipet/snipet/pkg/jsonx"
)

type IndexStatus string

const (
	IndexedStatusPending IndexStatus = "pending"
	IndexedStatusSyncing IndexStatus = "syncing"
	IndexedStatusIndexed IndexStatus = "indexed"
	IndexedStatusSkipped IndexStatus = "skipped"
	IndexedStatusError   IndexStatus = "error"
)

type IndexedKnowledgeItem struct {
	ID string `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`

	Hash      string        `gorm:"type:varchar(128);index" json:"hash"`
	IndexedAt *time.Time    `json:"indexed_at,omitempty"`
	Metadata  jsonx.JSONMap `gorm:"type:jsonb" json:"metadata"`
	Status    IndexStatus   `gorm:"type:varchar(20);not null;default:'pending'" json:"status"`
	Reason    *string       `gorm:"type:text" json:"reason,omitempty"`
	LastError *string       `gorm:"type:text" json:"last_error,omitempty"`

	IndexID         string  `gorm:"type:uuid;not null;index" json:"index_id"`
	KnowledgeItemID *string `gorm:"type:uuid;index" json:"knowledge_item_id,omitempty"`

	Index         *KnowledgeIndex `gorm:"foreignKey:IndexID;references:ID;constraint:OnDelete:CASCADE" json:"index"`
	KnowledgeItem *KnowledgeItem  `gorm:"foreignKey:KnowledgeItemID;references:ID;constraint:OnDelete:SET NULL" json:"knowledge_item"`
}

File: web/src/models/indexed-knowledge-item.ts
import { z } from "zod";

import { knowledgeIndexSchema, knowledgeItemSchema, type KnowledgeIndex, type KnowledgeItem } from "@/models/knowledge";

export const indexStatusSchema = z.enum(["pending", "syncing", "indexed", "skipped", "error"]);
export type IndexStatus = z.infer<typeof indexStatusSchema>;

export interface IndexedKnowledgeItem {
  id: string;
  hash: string;
  indexed_at?: Date;
  metadata: Record<string, unknown>;
  status: IndexStatus;
  reason?: string;
  last_error?: string;
  index_id: string;
  knowledge_item_id?: string;
  index?: KnowledgeIndex | null;
  knowledge_item?: KnowledgeItem | null;
}

export const indexedKnowledgeItemSchema: z.ZodType<IndexedKnowledgeItem> = z.lazy(() =>
  z
    .object({
      id: z.uuid(),
      hash: z.string(),
      indexed_at: z.coerce.date().optional(),
      metadata: z.record(z.string(), z.unknown()),
      status: indexStatusSchema,
      reason: z.string().optional(),
      last_error: z.string().optional(),
      index_id: z.uuid(),
      knowledge_item_id: z.uuid().optional(),
      index: knowledgeIndexSchema.nullable().optional(),
      knowledge_item: knowledgeItemSchema.nullable().optional(),
    })
    .strict(),
);
